### DISCUSSION TITLE
On evaluating the utility of some new features discussed in the meeting

### DISCUSSION THREAD

**Discussion by rivantsov** [graphql/graphql-wg#860] [General]
## Comments on some discussions at WG meeting on Jan 6 2022

Benjie - what a gorgeous CAT you have there! Too bad it looked like he didn't want to contribute, not a single 'meow' :). 

My comments concern discussions of 2 items: 'Named Lists' and 'Client controlled nullability'. Costs/benefits analysis for expressing extra type constraints like list elements nullability. I absolutely agree that we must evaluate potential utility against added complexity. What I am concerned with is how we approach this benefits evaluation. 

It seems to me, after listening and reading arguments, that we kinda assume that for an API the SDL schema doc (or introspection) is the ONLY information that will be provided to the consumer (human developer), so we need to express **as much details as possible like nullability of elements deep inside multi-level lists.** So the typical scenario is like this: the client programmer looks at schema doc, and starts programming the client based on what he/she sees in the doc. Thus the schema doc must provide as much details as possible about the types and fields and possible list contents, so there must be a syntax to express these extra constraints. 

However, I think we are making a mistake here, and as a result the utility/usefulness of a feature is overrated.

## 1. Semantics of the types/model express natural constraints
The real world types and fields are NOT meaningless foos/bars, they have explicit real-world semantics which make natural constraints clear and obvious in most cases. If we have an Invoice Type with Items list property, it is clear that this list cannot contain nulls, it does not make sense. On the other hand, if I have tic-tac-toe board represented as 2-dim array: 
  
    moves: [[String]]

then it is clear that individual cells are nullable, and all nulls is initial board. I am not saying that exact nullability notation is useless, but that its utility is limited, it just confirms what is already clear from semantics. 

The same goes for other constraints like suggested in 'NamedLists', for ex (quote): "we cannot represent collections with certain constraints.";  like "list contains at least 1 element". For an Invoice type as an example, it does not make sense to have invoice with 0 items. Thus 'non-empty' constraint is clearly derived from type semantics. The extra syntax expressing this non-empty constraint just confirms what is clear already. 

## 2. There is always documentation for non-trivial API which describes semantics and constraints

Any non-trivial API is delivered with documentation that describes the SEMANTICS of the API and its business semantics that is not expressible in syntax in the schema doc. Even if this documentation is only in extended description fields in the schema. It is these docs that contain all detail information about types and restrictions and validation rules for the data types and fields, in context of 'business flow'. And these docs should contain clear info on list like Invoice.Items - can it be null or can contain zero items.  It is not my wishful thinking that there must be docs - for non-trivial API the semantics and flows should be described somewhere, otherwise it is not possible to use the API. 

## 3. Tools may benefit?

So these 2 factors (semantics and extra API docs) make the utility (value added) of extra syntax for constraints quite limited and marginal in most cases, specifically for humans (devs) that can understand model semantics and can read docs. However, there is another type of information users - TOOLS, that cannot read plain docs and don't understand semantics of the API, they use introspection only. Can they use extra type constraint info?

During discussion of the features, there came a suggestion to find real world cases (lists specifically of higher ranks) to evaluate the potential benefits. What I suggest here, given the limited utility of the constraints for human developers (that I argued previously), we should instead look at tools like Graphiql and client generators, and ask a question - how it will benefit them, and indirectly humans that use them? Looking at tools is simpler I think than hunting for real world uses of some list types. 

And here I have doubts  - that these constraints will benefit the tools, but I might be wrong. 

## 4. Static constraints are overrated IMO
One more general consideration. In general, I think static type constraints (like null/nonNull) are overrated, and their possible impact on bugs in programming languages.  There is wide-spread belief that since many of runtime bugs are NullReferenceException, then if we tighten the control of null values, and add static constraints on fields and variables that value may not be null then we'll reduce number of these bugs. This is the reasoning behind recent introduction of new c# features of 'non-nullable reference types'. 

In my opinion, this is a false hope. 

The problem is - in many business cases the nullability of a field depends on the containing document 'Status'. A document (entity, complex type instance) goes thru a lifecycle; at the end, the document like Invoice MUST have certain fields nonNull - but only when it is in some final state like 'Approved'. Initially, however, it starts life in a Draft state, when many of the fields are null or empty lists, and it goes thru the process of filling things out. As a result, we CANNOT enforce a static constraint like NonNull on many fields, because they are NULL initially. 

The same goes for many other constraints. And many 'nullability' constraints depend on the invoice 'kind' or sub-type; there are many kinds, and each has it's own validation requirements. 

 *Added: the conundrum becomes very apparent when you design database table for Invoice and alike doc entities. You have to make many columns nullable even if it does not make sense for biz. It is in fact frustrating to see that you CANNOT use non-null constraint and have database assist you in validation/consistency - just because the doc starts in some DRAFT state for some short time where things can be null.*

From my experience, this is very common case, and it is THESE fields that are nullable initially, but should be assigned later, are the cause of NullReferenceExceptions down the road. These are real hard cases and hard bugs - the process expects Address to be filled but it's not. And static NonNull constraint cannot help, it cannot be used here. 

So what I am saying - don't sweat over expressing precise constraints on fields and values, they are NOT so useful after all. This is from my 25 years of experience in the industry, I cannot bring up exact numbers obviously. 

So to sum it up - if take all these considerations into account, the cost/benefit ratio of the features like 'NamedLists', or exact nullability of nested lists is quite LOW, at least it appears to me, given really non-trivial and potentially confusing syntax 

thank you

**Comment by twof** [comment]
Hi @rivantsov ! Thanks for your feedback. It's much appreciated. I think you're correct that semantic information is derivable by humans from API docs and field names, so I won't spend too much time addressing that. However, I want to clarify that the Client Controlled Nullability (CCN) proposal does not affect SDL. The syntax introduced by the proposal is solely used by developers writing operations.

 > 3. Tools may benefit?
 
While I think there is some human-readable semantic information provided by CCN, I don't think operations are read nearly as much as SDL (and API documentation) so I don't think it's nearly as important as what it enables for tooling. That's why that was the primary focus of the original RFC, so I'm glad you included a section about tooling.

I realized I didn't have the original RFC linked in the spec PR, so I've fixed that. Here's a link so you can read more about the motivations behind the proposal: https://github.com/graphql/graphql-wg/blob/main/rfcs/ClientControlledNullability.md

The TL;DR relevant to you is that at Yelp, code generation was rendered practically useless. Developers needed to write [wrapper types](https://github.com/graphql/graphql-wg/blob/main/rfcs/ClientControlledNullability.md#write-wrapper-types-that-null-check-fields) for every operation to handle all of the null-checking that was required. The Relay team saw similar problems at the company formerly known as Facebook, so they introduced [a Relay-specific version of CCN](https://relay.dev/blog/2021/12/08/introducing-the-new-relay-compiler/#what-does-the-new-compiler-enable) before we even started talking about CCN.

If you think those motivations aren't clearly translated over to the spec, let me know and I'll see what I can do to fix that :)

I've got some questions about the use case you described in "4. Static constraints are overrated IMO", but I'll have to come back to that later today.

**Reply by rivantsov** [reply]
thanks for reply. Pls note that I intended these comments to be about 2 RFCs - CCN and Named Lists. So some of them, apply more to one than to other. I will read the links you provided and maybe will have some more to say.

**Comment by rivantsov** [comment]
Ok, clearing the context that "It's all for client generators" - that changes the perspective. While I was listening to meeting recording, it all sounded like it's for human consumption. So my 'utility evaluation' ideas probably do not apply that much to CCN.

But then I have more questions, and even doubts. 

1. From explanations in docs and in PR to spec, it looks like the '?' marker is dependent on '!', it cannot appear by itself to change nullability, it can be used only as an extra to the '!' marker in the subtree. You never state this explicitly in the PR:
 https://github.com/graphql/graphql-spec/pull/895/files 
 
 but I think you should add this, for clarity. Otherwise it sounds like both markers are equal in their function and independent of each other. 

2. With '!' marker the client asks the server: *I know this field is nullable, but if it is null - please do some extra work for me; add an error message with the path to this field, and also nullify the containing object in the response. Yeah, I could do it here, but it's kind of annoying, lots of boring code, so I prefer you server do it for me.*
I am a server-side dev, and things like this - loading more work on the server - this triggers me. The server has a few CPUs to serve thousands of clients, while a client has the whole machine for itself. So for performance/thruput/server-load, it seems better to unload tasks from server, especially tasks that can be done on the client. 

3. Ok, now about '?'. With '!' the client asks the server to do some extra work. But with '?', it actually asks server to **violate its own API contract** : *hey, server, I know this field cannot be null in your schema, but this time pls return null here; just between me and you, nobody needs to know* That does not sound right to me. The contract should be respected - always.

4. So I can get OK with '!', but ? seems really problematic. So why not doing the following, using only ! (note: I am not client/js programmer, so I might say stupid things). Here is the scenario.

The client sends a request with ! on some nullable field. The server executes and ends up with null value for the field. The server then returns null for the field (it is nullable in schema!), but also adds an error to the response, with the path to the field. The client upon receiving response sees the error. So it needs to do the '?' part, nullify the containing object. I think it is not very complicated - take 'data' element, load into js object, follow the path of error (cut the last segment) and nullify it (parent object). Upon completion of all errors, the data object is serialized to json - and here you have it, the action of '?' done on the client. Then you can load this json into your data model. 

Again, not a js programmer, but - does this sound like something that can be done? If yes, then we can go with just ! modifier.

**Reply by twof** [reply]
I'm not a JS programmer either, for what it's worth. My background is Swift/iOS.

> From explanations in docs and in PR to spec, it looks like the '?' marker is dependent on '!', it cannot appear by itself to change nullability, it can be used only as an extra to the '!' marker in the subtree.

This is up in the air at the moment. Some folks have suggested that `?` should be necessarily paired with `!`, but we haven't decided that yet, which is why it's not reflected in the the spec PR. The current implementation has `?` entirely independent from `!`. `?` causes a field to become Nullable, and `!` causes a field to become Non-Nullable. Their only connection is incidental through the existing null propagation rules in the GraphQL spec which states that Non-Nullable fields will propagate null to their parent, halting propagation at the first Nullable parent. So `?` can halt null propagation caused by `!`, but they are otherwise entirely independent. Each can be used without the inclusion of the other.

I'm going to put up a discussion for the alternatives in a few and link it here so you can weigh in if you'd like.

> With '!' marker the client asks the server: I know this field is nullable, but if it is null - please do some extra work for me; add an error message with the path to this field, and also nullify the containing object in the response. Yeah, I could do it here, but it's kind of annoying, lots of boring code, so I prefer you server do it for me.

This is work that the server would already do if a field was defined as Non-Nullable in the schema, so I don't see this as a downside. Additionally, the spec only covers the behavior of the server, so if we'd like to standardize CCN, then it cannot dictate client behavior.

> The contract should be respected - always.

Quite a few other folks feel the same way as you, and the behavior of `?` is very much open to change.

**Reply by rivantsov** [reply]
> This is work that the server would already do if a field was defined as Non-Nullable in the schema

Well, the server would probably mine crypto for you IF it was a mining farm. But it isn't. If server did not promise something thru API schema, it is not reasonable to ask for it. "you would do it if you were..." - that logic does not work here.

**Reply by yaacovCR** [reply]
Actually, if I understand the flow correctly, a client side non-null causes the parent field completion to throw and therefore to appropriately quit working on any additional sibling fields. This is true in graphql-js even for async fields (although not in graphql-executor, see https://github.com/graphql/graphql-js/issues/2974 and https://github.com/yaacovCR/graphql-executor/pull/33)

A client side non-null can therefore be thought of as a client side execution instruction similar to skip/include, directing the flow of execution. It’s not _necessarily_ something the client can do herself.

**Reply by yaacovCR** [reply]
Although I don’t agree with the basic premise here, and agree with @twof response, the whole idea of non null fields are to offer null protection for clients, which is a convenience that the server offers in graphql. As of now, fields are null protected when defined as such within the schema on the server, and the idea of this PR is to allow clients to specify what null protection they want. It’s not really “extra” work, and it’s the same type of “extra” work that we already do, just doing it when actually necessary for the client…

**Reply by rivantsov** [reply]
I do not think analogy with skip/include works here, it is not the same. For skip, the goal is to reduce network traffic; the client can perfectly do it (throw away data), but it is waste of network resources, so asking server to filter it out makes sense. For '!' - there's no such excuse, it is simply loading work on server, work that can be done by client. 

On another subject
Can somebody explain what's wrong with my suggested solution for non-null but done purely on the client, without using '?' but with equivalent effect. Since this code can be put into the common client, it does not have to be in generated code for specific queries, it seems like it solves the problem. Copying it here: 
> The client sends a request with ! on some nullable field. The server executes and ends up with null value for the field. The server then returns null for the field (it is nullable in schema!), but also adds an error to the response, with the path to the field. The client upon receiving response sees the error. So it needs to do the '?' part, nullify the containing object. I think it is not very complicated - take 'data' element, load into js object, follow the path of error (cut the last segment) and nullify it (parent object). Upon completion of all errors, the data object is serialized to json - and here you have it, the action of '?' done on the client. Then you can load this json into your data model.
  
I would be probably ok with '!', but I feel against '?', it looks really confusing and unintuitive, and it is in fact ask to server to violate the contract (schema)

**Reply by twof** [reply]
@rivantsov We've discussed having a non-destructive alternative like you're suggesting (no change to `data`, error added to `error` with enough information for clients to do what they want) during some of the meetings. The feedback that I can recall was:
1. We can't dictate what clients do with responses. It's entirely up to them, so client behavior may diverge. This is a positive or negative depending on your perspective.
2. Naive clients (ie a one off bash script) will not benefit from CCN.

> the common client

What did you mean by this?

I'd also like to question the notion that propagating null to parent fields will be an undue burden on servers. If that were the case, I'd expect to hear that it's best practice to mark all fields nullable because doing otherwise causes a substantial performance hit. I've yet to hear anyone suggest that. If performance is a legitimate concern, however, we could consider adding performance tests to the GraphQL-JS implementation, but I think @yaacovCR is correct that the time potentially saved by not resolving, encoding, and transferring sibling fields is far far greater than the hit from a null propagation.

Even if the performance was worse, we'd be trading developer time for CPU time, and developer time is very nearly always more valuable.

**Reply by rivantsov** [reply]
Ok man, you lost me. I do not get the logic at all. Let me go thru it step by step. 
First, I assume that the 'solution' I outlined, with some extra code post-processing the response on the client and doing the '?' part - it would work. You can add this one code snippet once the processing pipeline of the response (on the client), and it solves the problem of nullifying containers of failed ! checks. 

What common client ? (you ask) - I meant that code like that can be either added to a JavaScript component that you use for sending/receiving GraphQL (like Apollo or Relay clients), or hook to processing event of client component. I assume that there's some place to put this code for the entire client app, that's it. 

In your first response on this thread, you explained that this feature will benefit code generators, here is quote: 
> ...at Yelp, code generation was rendered practically useless. Developers needed to write wrapper types for every operation to handle all of the null-checking that was required. The Relay team saw similar problems at the company formerly known as Facebook, so they introduced a Relay-specific version of CCN before we even started talking about CCN.. 

Your link to Relay compiler brings a post with examples of code with and without \@required directive. Ok, I understand this, looks like legit case. But... GraphQL spec is not a Santa's list, all good kids get whatever they want there. Only features are absolutely necessary, that **have no reasonably easy workarounds** - I hope you agree with that. 
Now in this case I see the situation like that: 

You: We need CCN
GraphQL WG (and me): ok, let's see.. how about this simple code executed on the client? (presents solution I described) - does it solve your problem? 
You: Yes (looks like)
GraphQL WG: you are welcome. Case closed, **this is an easy workaround**, we don't need CCN in spec

that's what is looks like so far to me. You presented problem for client code generators, we found a solution, no need to add anything to spec so far. 
> We can't dictate what clients do with responses

\- nobody dictates anything! you had a problem, we found a solution, without changing the spec. Use it if you want to.  

> naive clients (bash scripts) will not benefit from CCN 

should they? did they actually have problems? it was code generators for Relay who needed solution. 

Am I missing something?!

As for 'unique burden on the server". I do not say it will overload and crash servers, of course not. It's just feels the outcome of CCN is transferring some activity to Server, using some confusing syntax (it will be confusing and will require quite detail explanations with several sample scenarios); if will not crash the server, but as a server-side guy, I would prefer it the other way; move CPU work from server to client.

**Comment by benjie** [comment]
> Benjie - what a gorgeous CAT you have there! Too bad it looked like he didn't want to contribute, not a single 'meow' :).

Haha! :heart_eyes_cat: 

![image](https://user-images.githubusercontent.com/129910/149813382-0a5d2e8a-399e-4166-b1c9-bc9e702497a2.png)

**Comment by benjie** [comment]
> The problem is - in many business cases the nullability of a field depends on the containing document 'Status'. A document (entity, complex type instance) goes thru a lifecycle; at the end, the document like Invoice MUST have certain fields nonNull - but only when it is in some final state like 'Approved'. Initially, however, it starts life in a Draft state, when many of the fields are null or empty lists, and it goes thru the process of filling things out. As a result, we CANNOT enforce a static constraint like NonNull on many fields, because they are NULL initially.

You can encode that in GraphQL:

```graphql
interface Document {
  id: ID!
  title: String
  body: String
  # ...
}

type DraftDocument implements Document {
  id: ID!
  title: String
  body: String
  # ...
}

type PublishedDocument implements Document {
  id: ID!
  title: String!
  body: String!
  # ...
}
```

There's been some discussion over the years about adding literal values (e.g. `status: DocumentStatus.DRAFT`) to the GraphQL schema, but these haven't gone anywhere yet; in the mean time you could either add `status: String!` to all these, or you could use the `__typename` to derive the status.


> Added: the conundrum becomes very apparent when you design database table for Invoice and alike doc entities. You have to make many columns nullable even if it does not make sense for biz. It is in fact frustrating to see that you CANNOT use non-null constraint and have database assist you in validation/consistency - just because the doc starts in some DRAFT state for some short time where things can be null.

You can encode that in the database with check constraints:

```sql
create table documents (
  id int primary key,
  title text,
  body text,
  status text not null,
  check (status <> 'PUBLISHED' or (body is not null and status is not null))
);
```

This is, however, hard to extract meaning from programatically - you either need a human to interpret it or some complex code.

**Comment by rivantsov** [comment]
thanks for the response
First, note that my example with doc and its status - it is a very simplified example; real life situations are not so primitive, and much more messed up. Sometimes there's no explicit 'status' column, and constraint is dependent on multiple columns - and sometimes depends on data in OTHER tables. 

But the main problem with constraints like this - these are 'biz logic' rules. Which means they come from business/life. Which means they change. Get extended. And exceptions or special cases show up. Even if you succeed in encoding some flexible constraint, then pushed and made a Release - as soon as you're done, a biz-analyst/PM/customer shows up - "Ok, we need one special case, for this new  (contract/invoice/ins-policy) doc type, this X might be null, and this Y is never null." You have no power to say NO, it is business, that's how the process works in real world, and you have to change it, ASAP. It is a constant stream of changes and exceptions from rule, and a 'static' constraint is NOT a good way to express it.  

I know how to make database constraints, I spent a lot of time in relational databases. But if we learned anything in the last 20 years it is this: do NOT put biz logic into database, ever. Because logic will be changing non-stop and it will become nightmare to maintain and re-deploy.  And the sample conditional non-null constraint is biz logic, it comes from business. 

The same reasoning applies to your example in GraphQL. Expressing it in GraphQL - cool! but not actually helpful in real life, by the same reasoning as with databases. And looks to me the solution makes the model quite bloated, with extra type for each doc status. Introducing new status would require an extra type. 

The net conclusion again - **Static type constraints are overrated. Not useless, but not as useful as it might seem**. It's not worth inventing ways to express more complex type constraints, they won't be used much; just like fancy constraints in modern databases. 
  
PS Remember Eiffel progr language? the primary idea was contracts, static constraints, compile time validation. Did not work out. Life is too complex and changes too much to be put into static contracts.

**Reply by rivantsov** [reply]
Edit: changed 'static typing is overrated' to 'static type constraints are overrated', that expresses it better

**Comment by nalchevanidze** [comment]
# Topic: Named Lists

Thanks for opening the discussion. There are some points I agree with, but I differ on some.

## Documentation

> Thus 'non-empty' constraint is clearly derived from type semantics. The extra syntax expressing this non-empty constraint just confirms what is clear already (@rivantsov).

First of all, I don't think the restriction of the types behind the fields is always so obvious. For the invoice, this may be clear, but there are cases where it is not. Hypothetically, what if I never heard what the invoice is? (I know it's not realistic for the invoice, but it may make sense for another term in some domain).

## Plain documentation vs Types as documentation

> the client programmer looks at schema doc, and starts programming the client based on what he/she sees in the doc (@rivantsov)

Yes, that is correct. But types are also a way of documentation. The tedious thing about field documentation is its maintenance. When I update a certain behavior of invoices (e.g size), I have to simultaneously update the description of each field associated with the invoice in multiple parts. However we could define a list type with the name "Invoices" and its description, which we use for each field where we use invoices. this way we can also provide a description for all these fields.

```graphql
"""
some specific behaviour for invoices
"""
list Invoices
```

## Documentation can't be type-checked

> Any non-trivial API is delivered with documentation that describes the SEMANTICS of the API (@rivantsov).

However, the documents cannot be used by GraphQL client and server tools for static constraint checking ( as discussed in _Chapter 3._).

## Static constraints are not overrated

> One more general consideration. In general, I think static type constraints (like null/nonNull) are overrated (@rivantsov).

I see it a different way. As developers, we want to ensure that errors we can avoid at compile time are handled by the type system, not by us. Ideally, the program is a formal proof of our requirements (see Curry-Howard correspondence). However, if we specify an incorrect formula, our program will give the false impression of proof of correctness. In other words, our application will crash at some point in the runtime (in the worst case, in production).

> __Curry-Howard correspondence__: The type of a program can be understood as a formula and the program itself corresponds to its logical proof ([Samuel, P. 12
](https://www.lix.polytechnique.fr/Labo/Samuel.Mimram/teaching/INF551/course.pdf),[Michael R](https://cs3110.github.io/textbook/chapters/adv/curry-howard.html#the-correspondence))

In your example, I see that documents with the status "DRAFT" and "PUBLISHED" have different characteristics and should not be unified as one type instead of having two different types as @benjie represented.

## Use cases and Client benefits

> we should instead look at tools like Graphiql and client generators, and ask a question - how it will benefit them, and indirectly humans that use them? Looking at tools is simpler I think than hunting for real world uses of some list types (@rivantsov).

First, this feature is beneficial not only for the clients but also for the servers since the named list automatically validates all list values that the server sends to the client. Second, some common types like (`Set` `Map`, n sized `Vector`) are supported by almost every language. So these languages will benefit from `named lists` to represent those. There is also the type `NonEmpty` in Haskell with an additional benefit.

**Reply by rivantsov** [reply]
Documentation
(about mentally deriving constraints from API semantics described in Docs). You write: 

> Hypothetically, what if I never heard what the invoice is? (I know it's not realistic for the invoice, but it may make sense for another term in some domain).

In any domain, you should know something about domain area, if you are doing UI or a client for a GraphQL API. But in any case, I do not mean semantics always give you all and you do not look at type constraints. But very often the semantics, the meaning helps, especially with complex constraints. And thus the extra benefit of constraints or extra types like Set or NonEmpty list is diminished. We are in cost/benefit analysis here.

Documentation maintenance
> However we could define a list type with the name "Invoices" and its description, which we use for each field where we use invoices. this way we can also provide a description for all these fields. 

You can do exactly this today, with a Type with a single list field: 

type Invoices {
	list: Invoice[] 
}

> However, the documents cannot be used by GraphQL client and server tools for static constraint checking 

I assume you think of a scenario like this. On a server, there is some Response Verifier method that sits on the output of the server and verifies every response before it is sent to client. It verifies the constraints from API Schema like notNull for fields. Well, as a server programmer, I can tell you - there is no such thing. For two reasons: 
1. First, overwelming majority of responses would be OK, so validating them is a pure waste of CPU cycles
2. Secondly, if I get a null for nonNull field, it means something when horribly wrong in my data processing logic, and it should be caught and an exception will be bubbling up the call stack. So you will not get null, you will get an exception; and top exc handler will build error response. 

Hence auto-verifier is useless. At least on the server. If you need verifier to catch null values on non-null fields, then there is something really wrong with your code. You'd better fix it, instead of looking for verifiers.  

Not sure about clients, running verifier of request; but it seems useless too. If you send invalid request, you will get error from server. What's the difference with client side verifier error? Nothing. As long as you're not sending too many bad requests, you're fine, and server will return errors for you. 

Tools like Graphiql. Well maybe they can show some squiggly line in some place, to indicate constraint violation, but that's all. If I write invalid request and click execute, Graphiql or any other tool will execute it - and it should! Simply because tool's validation code should never override user's decision

_about Curry-Howard and stuff - to be continued_

**Comment by rivantsov** [comment]
I have posted a feedback on original NamedLists thread: 
https://github.com/graphql/graphql-spec/issues/914#issuecomment-1020865156

**Reply by rivantsov** [reply]
About proof of correctness and Haskell and all that stuff. 
> as developers, we want to ensure that errors we can avoid at compile time are handled by the type system, not by us. Ideally, the program is a formal proof of our requirements (see Curry-Howard correspondence).

well... here comes again, the dream of a magic verifier of correctness. And you mentioned Haskell.
Each time this stuff comes up - I bring up these link: 

https://jaxenter.com/disadvantages-of-purely-functional-programming-126776.html

in the post, the link about quick sort is dead, but here is an updated URL: 
https://sudonull.com/post/75314-Parallel-quick-sort-on-Haskell-and-how-hard-it-was-to-write 

So, please read the quick sort story in Haskell. Multiple attempts by PhD's, experts in Haskell, several years... and multiple failures, published code contained bugs. Quick sort, not a super-high math. 

My question - where were the correctness verifiers-proofers. Wasn't Haskell created for this kind of stuff? With all the advance nature of Haskell type system, why couldn't they quickly produce correct code?!

What I am saying is this. Practical programming of real life systems and this Curry-Howard academic stuff - these are 2 complete separate universes. All this proof of correctness theoretical stuff simply DOES NOT WORK in real life. Useless. Tried and failed many times. So let's leave it alone, all these proofers and hopes that advanced enough type system (a-la category theory?!) will save us all some day.

Also, let's remember: **GraphQL is not a programming language**. It is a language to express queries and (relatively simple) data manipulation instructions, from client to server. Not for coding complex algorithms.  So analogies with programming languages - might work but to a very limited extent. 

My main line again: inventing more complex type system has limited benefit in GraphQL.

**Reply by nalchevanidze** [reply]
Thanks for reply :) I am sorry if my response left a different impression than I intended. I have no illusions that the type system will eliminate all possible bugs. However, if it eliminates about 95% of them, we can deal with the last 5%. Besides, the problem described in the article was quite interesting in terms of performance and memory issues. However, it is not completely relevant as a counterargument for the type system, since the problem is caused by the pure nature of the language, but not by the complexity of the Haskell type system. Besides, if fewer types are better suited for real-world use cases, then it's hard to explain why most people prefer Typescript (i.e. in GraphQL.js) over JS.

Haskell handles side effects with monadic IO functions. These monadic functions cannot guarantee execution safety any more than procedural or OOP language. Therefore, classical mutation errors may occur at runtime. Since the type system can only guarantee that if we expect apples, we will not get oranges at runtime. Besides, the use of a language always depends on the domain. Therefore, evaluating the uselessness of a language is not an obvious question. Although there is an opinion that Haskell can be useless in real life. In the last decades, more and more companies have used Haskell to solve their domain-specific problems. The examples that immediately come to mind are Haxl and Cardano.

GraphQL is not a programming language, however, it gives us guidelines for modeling applications that use it. Therefore, the absence of specific features in the type system implies limitations in the target language.

**Reply by rivantsov** [reply]
Well, you brought up this stuff, quote: " Ideally, the program is a formal proof of our requirements (see Curry-Howard correspondence)."
the story about quick sort in haskell - is my counter example about absolute failure of this formal proofing concepts; PhD's publishing solutions with subtle bugs - where are correctness proofers?! my point is - do not bring this stuff up in places like here, you won't be taken seriously, and you get a sarcastic questions like 'oh - did they finally figure out the quick sort?!'
let's leave this theoretical stuff behind, and deal with practical questions, meaning - benefits vs cost (over complexity and possible confusion)

**Reply by nalchevanidze** [reply]
Alright :)

Please note that I started the sentence with the word "Ideally". Nothing is ideal in the real world of programming, but we always have to deal with some kind of trade-offs. however, that doesn't mean we shouldn't try to achieve it to some degree.

also, let's not forget that Haskell is a programming language, not a proof assistant. Haskell guarantees a high level of safety (but still not 100%) when writing pure code. if we violate this (which is what happened in the article), we are almost working with a regular language and the "correctness proof" is pretty irrelevant that case.

Yes. I also think we should get back to the original topic of `Named Lists` (I saw your comment on them and will respond later).
